AnyReport alternative, Report<dyn Context> that can be defaulted to with error_stack::Result<T>

zakstucke opened the pull request:
Alternative to #7256 that adds `?` support to `Report<dyn Context + Send + Sync + 'static>` instead of an `AnyReport` wrapper.

This is even more powerful because the `error_stack::Result` type alias can be updated to: 
```rust
pub type Result<T, C = dyn Context + Send + Sync + 'static> = core::result::Result<T, Report<C>>;
```

Which makes switching between specific contexts and a catch all completely painless. `Result<T, E>` can simply remove the `E` to make it catch-all.

This is at the cost of a breaking change, the new `From` impl breaks `Report::from` type inference, meaning some `Report::from` need changing to `Report::new`. #7256 is a breaking change.

zakstucke commented:
I have now updated the `Report` struct definition to:
```rust
struct Report<C: ?Sized = dyn Context + Send + Sync + 'static>
```
this makes an "untyped report" a first class citizen and means `Report` and `Report<C>` can be used interchangeably, an untyped `Report` acts like `AnyReport` does in #7256.

The concept of an untyped `Report` feeds nicely into a new feature, owned type extraction:
```rust
impl<C> Report<C> {
    pub fn pop_current_context(self) -> (Report, C)
        where
            C: Send + Sync + 'static
}

impl<C: ?Sized> Report<C> {
    pub fn downcast<T: Send + Sync + 'static>(self) -> Result<T, Self>

    pub fn downcast_take<T: Send + Sync + 'static>(self) -> Result<(Report, T), Self>
}
```

This addresses the second painpoint of ES, some error types don't implement `Clone`, or cloning is very expensive, despite needing to own them.

This PR now implements these methods, making sure to not affect the printable representation of a report despite extracting types, by eagerly formatting on extraction.

TimDiekmann commented on libs/error-stack/src/report.rs:
````diff
@@ -452,6 +452,89 @@ impl<C> Report<C> {
+    /// let (_report, io_error) = report.pop_current_context();
+    /// assert_eq!(io_error.kind(), io::ErrorKind::NotFound);
+    /// ```
+    pub fn pop_current_context(self) -> (Report, C)
````
This is actually an interesting use-case, however, I'd probably revert the return statements.

zakstucke replied:
I agree! Updated

TimDiekmann reviewed (commented):
As mentioned by @vilkinsons, we're going to discuss this -- most likely on Tuesday (I'm currently on vacation). I like `Report<dyn Context>` much more than the artificial error type. But a few comments regardless:

On `Result<T, C = Report<dyn Context>>` and the deprecation: We deprecated `Result` for a quite simple reason: `Result<T, E>` behaves differently as soon as you import `error_stack::Result`. This was quite annoying and -- as you mentioned -- this can be done downstream quite easily. If it would be defined like this
```rust
type Result<T, C = dyn Context> = Result<T, Report<C>>
```
it would still behave differently from `std::result::Result`, so I would not touch `Result`.

On `dyn Context` in general: `Context` is deprecated in favor of `core::error::Error`, which was stabilized a while back. This means, `Context` is not needed at all and we should not build anything on top of it but simply use `dyn Error` here.

On defaulting `C` in `Report<C>` to `dyn Context`: If we would add `dyn Context` as possible target, I don't like us to add a default to it as this encourages omitting `C` in most cases. The user should explicitly opt-in to the generic error. `Report<dyn Context>` (or dyn Error) should only be used if there really is not another way and a named `C` should be used per default.

zakstucke commented:
> ...behaves differently as soon as you import error_stack::Result. This was quite annoying...

I agree, I actually don't use this, but a renamed `ResultStack` version instead that does the same thing, the alias is useful, it just shouldn't overwrite `core::result::Result`. Anyhow, it's scheduled for removal and can be done in userland, I've reverted the change.

> Context is not needed at all and we should not build anything on top of it but simply use dyn Error here.

Agreed, updated!

> I don't like us to add a default to it as this encourages omitting C in most cases. The user should explicitly opt-in to the generic error.

I disagree with this, but understand the stance. I think having a "split"/"untyped" `Report` as a concept is powerful and useful as shown in the `C` extraction methods, I also feel ES can explain in docs and style how it's intended to be used, without introducing hurdles for users that do want to use ES in more flexible ways. Like with the alias, it makes migrations easier and more gradual.
Given a user can re-type alias `Report` in userland, I'd be fine with removing this default if it unblocked this PR.

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -243,7 +243,7 @@ use crate::{
 #[must_use]
 #[expect(clippy::field_scoped_visibility_modifiers)]
-pub struct Report<C: ?Sized> {
+pub struct Report<C: ?Sized = dyn Error + Send + Sync + 'static> {
```
We don’t want to encourage people to use `dyn Error` as the common case. People should explicitly specify, if a `Report` is untyped. 
```suggestion
pub struct Report<C: ?Sized>
```
Also, we should add a documentation in which scenarios `Report<dyn Error>` is discouraged and when it’s useful:
- in return statements in an `Err` variant is discouraged, as a typed error is more representative. `Report<dyn Error>` does not add any contextual information to the `Report` which is a core principal of `error-stack`
- in argument position (e.g. for error handling) might be useful if the handling code does not need to know the current context. For example if it just searches for a `Suggestion` in the `Report` it doesn’t need to know `C`

The last bullet brings me to another point: We might need to make sure, that `current_context` is not callable on `Report<dyn Error>`. I haven’t looked at this, yet, but it might cause issues. Maybe it also just works.

TimDiekmann replied:
Also, the changelog needs updating

zakstucke replied:
> Yes, it was allowed but it was not possible to construct it (unless you use unsafe transmutation). I'd like to have some documentation around dyn Error before merging this.

Got it, cool! I've added a changelog entry, do you want to take a stab at documentation? I think the you have a much better feel for how you'd like it presented.

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -471,6 +554,24 @@ impl<C> Report<C> {
     }
 }
 
+impl<E: Into<Report<E>>> From<E> for Report<dyn Error + Send + Sync + 'static> {
```
`dyn Error` is only a marker type, it does not really serve any purpose. In theory, it could be anything, such as `()`. `dyn Error` is descriptive and `Report<C>` always implements `Send + Sync` as it only can be constructed with such a type. It should be enough to just use
```suggestion
impl<E: Into<Report<E>>> From<E> for Report<dyn Error> {
```
This also aligns better with not setting a default for `C` as it’s less to type.

zakstucke replied:
> In theory, it could be anything, such as ().
It couldn't be () as that's `Sized` and you'd lose the cool ability to specialise methods that don't apply to `dyn Error`.

You're right on the `Send + Sync + 'static`, cool! Updated.

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -471,6 +554,24 @@ impl<C> Report<C> {
+    }
+}
+
+impl<C> From<Report<C>> for Report<dyn Error + Send + Sync + 'static> {
```
```suggestion
impl<C> From<Report<C>> for Report<dyn Error> {
```

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -471,6 +554,24 @@ impl<C> Report<C> {
+            frames: value.frames,
+            _context: PhantomData,
+        }
+        .attach_printable(format!("into untyped Report at {}", Location::caller()))
```
It’s a good idea to add some attachment to cover the location, but I don’t know if we want to do that using a `String` or just append `Location::caller`. @indietyp thoughts?

zakstucke replied:
It would be good to specify what's happened here, given no default now I guess that would be `"into Report<dyn Error> at {}"` instead?

zakstucke replied:
Makes sense! Will update

TimDiekmann commented on libs/error-stack/src/report.rs:
````diff
@@ -452,6 +452,89 @@ impl<C> Report<C> {
+    /// let (io_error, _report) = report.pop_current_context();
+    /// assert_eq!(io_error.kind(), io::ErrorKind::NotFound);
+    /// ```
+    pub fn pop_current_context(self) -> (C, Report)
````
I see the motivation behind this function, but I’m having my issues with that.
The encoded behavior is weird. `pop` typically removes something from a stack, but here it’s replaced with a dummy value leaving the `Report` in another behavior state. We might want to think about another function, such as `replace_current_context` to achieve this, but `pop` should really remove a context. However, removing a context comes with various other issues:
- What happens if we hit a multi-error after popping a context? I don’t see us supporting `Report<[dyn Context]>` (at least not initially). `[dyn Context]` first of all is not valid Rust, as slices must not have unsized types as values. Also, it effectively says, that the multi-context could be of a different type, which really is against the concept of multi-contexts.
- What happens if we hit the last context? We would need to return an `Option<Report>` instead

It would be good if we would not add such a functionality as drive-by.

TimDiekmann commented on libs/error-stack/src/report.rs:
````diff
@@ -452,6 +452,89 @@ impl<C> Report<C> {
+    /// assert_eq!(io_error.kind(), io::ErrorKind::NotFound);
+    /// ```
+    #[must_use]
+    pub fn into_current_context(self) -> C
````
This discards everything from the current context the `Report`. This includes other contexts and all attachments. Ideally, this would somehow return the rest of the `Report` somehow as well (similar to what I would expect `pop_current_context` to behave), but sometimes the sole context is everything which is required. Do you have a concrete use-case where you cannot use `current_context()` and need `into_current_context`?

It would be good, if we would not add this functionality as a drive-by.

zakstucke replied:
> Ideally, this would somehow return the rest of the Report somehow as well (similar to what I would expect pop_current_context to behave)

This is exactly what `pop_current_context()` does from a user's perspective, if you don't want the top level context, you can use `downcast_take()`. 

>  Do you have a concrete use-case where you cannot use current_context() and need into_current_context?

Yes, some external error types do not implement `Clone`, but we need to extract the type to handle it, after logging the report. **This has been the most painful part of a migration to ES as there is no workaround currently.**, likewise a lack of `downcast/downcast_take`.

This exists on top of `pop_current_context()`, because it's slightly more efficient - it doesn't need to eagerly extract fmt strings like the pop variant does.

indietyp replied:
couldn't just just `downcast_mut`, then `mem::take`? This wouldn't violate the structure and you could just replace it with a dummy variant.

We cannot just _simply_ replace the `Context` with an attachment, it _completely_ changes the semantics of a report *and* violates many of the internal variants. Each attachment is attached to a context, by replacing it with another context you just "joink" the rug and attach attachments to an unrelated context that might not even find these useful. What about the types it provides via the provide api?

1. Every context *must have* a single context as the first layer, both formatting and internals rely on it. It's a context stacked on top of other contexts with attachments sandwiched in between
2. As @TimDiekmann eluded to there are some questions about `Report<[C]>` types, how do they fit into the equation - when I replace a report, does that mean that one arm now doesn't have a report? What about the introduced asymmetry? Which arm is chosen?

Also remember: we're talking about the unhappy path - do we really need to micro-optimize for the sake of it *without any measured benchmarking*, if so - _how can we implement this without sacrificing **correctness**_?

TimDiekmann commented on libs/error-stack/src/frame/mod.rs:
```diff
@@ -90,6 +92,72 @@ impl Frame {
+    ///     can be set to false when the `Report` will no longer be used to avoid the cost of
+    ///     formatting
+    #[must_use]
+    pub(crate) fn downcast_take<T: Send + Sync + 'static>(
```
As mentioned above, we should not add this as part of this PR. This and any other related work should be removed from this PR to make this PR focused on `Report<dyn Error>`.

Also, the behavior to replace the context with another opaque context feels like an anti pattern. If we want such a functionality, the new context should be part of the signature, and this could be defined on any `Report` (e.g. `report.replace_context(MyContext)`. Let's remove it for now.

TimDiekmann reviewed (changes requested):
Hey @zakstucke and thank you for your patience.

We agreed that we’re going to add `Report<dyn Error>`. However, we see a few issues in the current implementation. Mainly, this boils down to not adding the consuming operations (`pop_current_context` and `into_current_context`), at least not yet.
I’d propose we close #7256, and focus this PR on only adding the code for `Report<dyn Error>`. As an alternative, I’d be happy to implement it on my own PR (either later this day or sometimes this week), that’s up to you.
For the future if we want to add more current-context related API, we should park that under the `unstable` feature flag.

Another thing we might want to explore is if we could solve #4295 with `Report<dyn Error>`.

TimDiekmann commented on libs/error-stack/tests/test_conversion.rs:
```diff
@@ -88,7 +88,7 @@ fn error() {
 #[test]
 fn into_report() {
-    let report = io_error().map_err(Report::from).expect_err("not an error");
+    let report: Report<io::Error> = io_error().map_err(Report::from).expect_err("not an error");
```
This change is very concerning and after running the full CI (which uses pretty much all features of error-stack), this is a quite big breaking change and a big QoL regression: https://github.com/hashintel/hash/actions/runs/15422334171/job/43400766905?pr=7274

zakstucke replied:
This specific update was just done like this to maintain the under-test semantics of that specific test, otherwise I would've just replaced `Report::from` with `Report::new`. 

The other CI failures mainly seem to be about needing `impl Into<Report<C>>` -> `impl IntoReport<Context = C>` which I think is the new recommended way of defining these anyway. I hope at least! I'll try updating them all.

zakstucke replied:
@TimDiekmann lets see how CI likes it now 🙏 , each of the changes I made repo wide made sense to me and improved things or stayed the same don't think a blocker.

TimDiekmann reviewed (changes requested):
Ii think we can make the implementation less breaking. To require specifying the `Report` type is suboptimal. 

Also, I like to have a bit more test coverage.
The `size` test can be expanded with this
```diff
     #[test]
     fn size() {
         assert_eq!(mem::size_of::<Report<()>>(), mem::size_of::<*const ()>());
+        assert_eq!(
+            mem::size_of::<Report<dyn Error>>(),
+            mem::size_of::<*const ()>()
+        );
     }
```

and some invariants should be checked:
```diff
   const fn report() {
         assert_send::<Report<()>>();
         assert_sync::<Report<()>>();
         assert_static::<Report<()>>();
+
+        assert_send::<Report<dyn Error>>();
+        assert_sync::<Report<dyn Error>>();
+        assert_static::<Report<dyn Error>>();
    }
```

Two notable things:
We can provide a `current_context` method for `Report<dyn Error>` (but I would probably prefer to add that in a later PR):
```rust
impl Report<dyn Error> {
    #[must_use]
    pub fn current_context(&self) -> &dyn Context {
        self.frames()
            .find_map(|frame| match frame.kind() {
                FrameKind::Context(context) => Some(context),
                FrameKind::Attachment(_) => None,
            })
            .unwrap_or_else(|| {
                // Panics if there isn't an attached context which matches `T`. As it's not possible
                // to create a `Report` without a valid context and this method can
                // only be called when `T` is a valid context, it's guaranteed that
                // the context is available.
                unreachable!(
                "Report does not contain a context. This is considered a bug and should be \
                reported to https://github.com/hashintel/hash/issues/new/choose"
            );
            })
    }
}
```

And your assumption about when exactly what signature is used is not fully correct.
We have three different states of `Report`:
1. `Report<C>`
2. `Report<dyn Error>`
3. `Report<[C]>`

To share some implementations, these signatures can be used

|                                     | `Report<C>` | `Report<dyn Error>` | `Report<[C]>` |
|:-----------------------------------:|:-----------:|:-------------------:|:-------------:|
|         `impl<C> Report<C>`         |      ✅      |          ❌          |       ❌       |
| `impl Report<dyn Error>` |      ❌      |          ✅          |       ❌       |
| `impl<C> Report<[C]>` |      ❌      |          ❌          |       ✅       |
| `impl<C: ?Sized + Error> Report<C>` |      ✅      |          ✅          |       ❌       |
|     `impl<C: ?Sized> Report<C>`     |      ✅      |          ✅          |       ✅       |

This might be useful when implementing functionality on the different states.

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -471,6 +471,27 @@ impl<C> Report<C> {
     }
 }
 
+impl<E: Into<Report<E>>> From<E> for Report<dyn Error> {
+    #[track_caller]
+    fn from(value: E) -> Self {
+        value.into().into()
+    }
+}
+
```
I don't know if we want to revert the other changes, but it would be possible. As an alternative implementation we could also use `impl<E: Context> From<E>`.

TimDiekmann replied:
Regardless of the implementation, we should not use the `.into()` below but construct it ourselves. Otherwise, we'd have the attachment that it got converted to `Report<dyn Error>`.

zakstucke replied:
Fixed the `.into().into()`.

We could change to `impl<E: Context> From<E>`, it doesn't affect the type inference tweak as far as I can tell.

> This is actually the issue why the change is largely breaking. Generally, a double .into() is a sign that this will easily result in compile errors if no type is specified.

Unfortunately it's the specific part I opened the PR for, if you look at the actual changes required across your repo, and the additional benefit of allowing `impl<C: Context> From<C> for Report<[C]>`, I hope you'll be fine accepting this.

zakstucke replied:
> I thought the motivation for the pull request was to allow an intermediate report?

The title of the PR:
> Alternative to https://github.com/hashintel/hash/pull/7256 that adds ? support to Report<dyn Context + Send + Sync + 'static> instead of an AnyReport wrapper.

Would you be open to having this impl behind a feature flag? The tradeoff here i would make always, but understand completely why you wouldn't. It really was the crux of the PR for me, motivation explained back in #7256. 

The PR would still allow a user to reinvent a wrapping report type that has this dyn Error inside but would be much nicer if i could enable a feature and not reinvent the wheel on top

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -471,6 +471,27 @@ impl<C> Report<C> {
+            frames: value.frames,
+            _context: PhantomData,
+        }
+        .attach_printable(alloc::format!(
+            "into Report<dyn Error> at {}",
+            Location::caller()
+        ))
```
You mentioned you updated this, but it's still a string. Either, we want to have a dedicated attachment struct which wraps the location or we simply add the `Location`. The latter is easier to modify when outputting, which is why I would prefer that route (people frequently asked how to hide the location, and this is a potential issue as this message is not easy to hide if it's a string or a private struct).

zakstucke replied:
> Makes sense! Will update

Sorry, just hadn't got to this one yet! Update**d**

TimDiekmann commented on libs/error-stack/src/report.rs:
```diff
@@ -892,6 +913,12 @@ impl<C> From<Report<C>> for Report<[C]> {
     }
 }
 
+impl<C: Context> From<C> for Report<[C]> {
+    fn from(ctx: C) -> Self {
+        Report::new(ctx).into()
+    }
+}
+
```
cc @indietyp thoughts? I think it's fine but probably should be generalized using `into_report()` using
```rust
impl<E: IntoReport<Context: Sized>> From<E> for Report<[E::Context]> {
    fn from(error: E) -> Self {
        Self {
            frames: error.into_report().frames,
            _context: PhantomData,
        }
    }
}
```

but only if we keep `impl<E: Into<Report<E>>> From<E> for Report<dyn Error>` (or however we'd implement it)

zakstucke replied:
This generalisation doesn't quite work: 
```
conflicting implementations of trait `std::convert::From<report::Report<_>>` for type `report::Report<[_]>`
conflicting implementation for `report::Report<[_]>`
```

TimDiekmann commented on libs/error-stack/CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@ All notable changes to `error-stack` will be documented in this file.
 - `Extend` is no longer implemented by `Report<C>`, instead it is implemented on `Report<[C]>`, either use `From` or `Report::expand` to convert between `Report<C>` into `Report<[C]>`. ([#5047](https://github.com/hashintel/hash/pull/5047))
 - `extend_one` has been renamed to `push` and is only implemented on `Report<[C]>`. ([#5047](https://github.com/hashintel/hash/pull/5047))
 - `bail!(report,)` has been removed, one must now use `bail!(report)`. This is in preparation for the unstable `bail!` macro that allows to construct `Report<[C]>`. ([#5047](https://github.com/hashintel/hash/pull/5047))
+- `From<C> for Report<[C]>` and support for constructing `Report<dyn Error>` with `From<E: IntoReport<E>>` and `From<E: Report<C>>` alter type inferrence in some cases. Each instance will require one of: replacing `Report::from` with `Report::new`, replacing `Into<Report<C>>` with `IntoReport<Context = C>`, or specifying a concrete `C`. This makes multi-error reports easier to instanstiate and al ...
```
Fundamental parts are missing:
- We did introduce `Report<dyn Error>`. At the same time, it should be said, it's discouraged to be used as return type and is mainly interesting for locations where we don't need the context type, e.g. in error-handling. It's expected to be seen more in argument position in general
- I said above, that we're not going to add the additional `From` implementation involving `into().into()` as it breaks inference. It's not expected, that `Report::from` or `Into<Report<C>>` needs to change.
- I don't see, why we added a `From` implementation for multi-errors. Even if we would, this PR has a fundamentally other focus.

zakstucke replied:
Updated

zakstucke commented on libs/@local/repo-chores/rust/src/benches/analyze/tracing.rs:
```diff
@@ -41,7 +41,7 @@ impl FoldedStacks {
         &self,
         mut options: flamegraph::Options,
-    ) -> Result<FlameGraph, Report<impl Error + Send + Sync + 'static>> {
+    ) -> Result<FlameGraph, Report<dyn Error>> {
```
Type inference change from the `From<E>` impl prevents `impl Error`, will revert once the from impl is removed / feature flagged
